interface/language/csv/translation_utilities.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+/* Copyright (C) 2014 Kevin Yeh <kevin.y@integralemr.com>
+ *
+ * LICENSE: This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://opensource.org/licenses/gpl-license.php>;.
+ *
+ * @package OpenEMR
+ * @author  Kevin Yeh <kevin.y@integralemr.com>
+ * @link    http://www.open-emr.org
+ */
+
+function find_or_create_constant($constant)
+{
+    $sqlFind = " SELECT cons_id , constant_name FROM lang_constants where BINARY constant_name = ?";
+    $result = sqlStatement($sqlFind, array($constant));
+    if ($result) {
+        $row_count = sqlNumRows($result);
+        if ($row_count == 1) {
+            $row = sqlFetchArray($result);
+            return $row['cons_id'];
+        }
+        if ($row_count > 1) {
+            error_log("Duplicate Entries for language constant:" . $constant);
+            $row = sqlFetchArray($result);
+            $retval = $row['cons_id'];
+            while ($row = sqlFetchArray($result)) {
+                $sqlDelete = " DELETE FROM lang_constants where cons_id = ? ";
+                sqlStatement($sqlDelete, array($row['cons_id']));
+                $sqlDelete = " DELETE FROM lang_definitions where cons_id = ? ";
+                sqlStatement($sqlDelete, array($row['cons_id']));
+                error_log("DELETED Definitions for duplicate constant:" . $constant . "|" . $row['cons_id']);
+            }
+            return $retval;
+        }
+        if ($row_count == 0) {
+            $sqlInsert = " INSERT INTO lang_constants (constant_name) VALUES (?)";
+            $new_index = sqlInsert($sqlInsert, array($constant));
+            return $new_index;
+        }
+    }
+}
+
+function update_metainfo($constant, $definition, $source, $set_source = false)
+{
+    if ($source != "") {
+        $sqlUpdate = " UPDATE ippf_lang_definitions set " . $source . "= ? where BINARY constant_name = ?";
+        $parameters = array($definition, $constant);
+        sqlStatement($sqlUpdate, $parameters);
+        if ($set_source) {
+            $sqlUpdate = " UPDATE ippf_lang_definitions set source = ? where BINARY constant_name = ?";
+            $parameters = array($source, $constant);
+            sqlStatement($sqlUpdate, $parameters);
+        }
+    }
+}
+
+function verify_translation($constant, $definition, $language, $replace = true, $source = "", $set_metainfo = false, $preview = false)
+{
```

Comment:
when add/update stuff also need to get this into the lang_custom table via the insert_language_log() function. I wouldn't worry about the $row count > 1 scenario since this should never happen and would get sorted out if the user ever ran through the Manage Translations process

Reply:
What if the spreadsheet is distributed as "official" translations and are not considered customizations? That's the normal use case here.

Reply:
I think that will be ok and will then just mean that a user could use IPPF official translations via this csv method, which will then take precedence and carry over even if a user downloads the "official" openemr translations followed by the Manage Translations process (Manage Translations is basically adding back all the translation work that the instance had done locally, which would include the csv method in also).

Reply:
actually, let me think about this more before doing work on this part

Reply:
quick question:
Are these IPPF instances doing their own custom local translations on top of the csv set (ie. will they have local things that need to override/survive the csv set)?

Reply:
Yes it's likely that some sites will. Probably need to preserve their custom translations.

Reply:
Maybe a checkbox to specify which way to go?

Reply:
Interesting issue.
1. Current Manage Translations allows local instances keep their local changes after importing an official set. So in IPPF's case, won't want to write to lang_custom when do this (since that instance can do the official IPPF import and then run Manage Translations to bring back their local changes).
2. I'm just not sure what will happen in the wild in the future. I think actually best to not write the csv import to lang_custom and not even offer an option (since it will just be confusing). If really necessary and wanted to have it all they could do (ie. by not adding to lang_custom just adds one more step and keeps it less confusing):
- make their local custom changes
- bring in openemr's official translations
- bring in csv translations
- run Manage Translations (this will bring back their local custom changes)

So, would just keep it how you have it :)
And if address the other minor issues and testing well, then feel free to bring it in.